File: 43162_2.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<vector<int>> adjacencyMatrix;
vector<int> vertices;
vector<int> stack;
int networks = 0;

void DFS(int vertex);

int solution(int n, vector<vector<int>> computers) {
    int answer = n;
    
    adjacencyMatrix = computers;
    vertices = vector<int>(n, 500);
    
    for(int i = 0; i < n; i++) {
        if(vertices[i] == 500) {
            networks += 1;
            vertices[i] = networks;
        }
        
        DFS(i);
    }
    
    return networks;
}

void DFS(int vertex) {
    for(int i = 0; i < adjacencyMatrix[vertex].size(); i++) {
        if(adjacencyMatrix[vertex][i] && vertices[i] == 500) {
            vertices[i] = vertices[vertex];
            DFS(i);
        }
    }
}

File: 43162_2.swift
import Foundation

func solution(_ n:Int, _ computers:[[Int]]) -> Int {
    var numberOfNetworks = 0
    var networks = Array(repeating: 0, count: n)
    for computer in 0..<n { dfs(computer, &numberOfNetworks, &networks, n, computers) }
    return numberOfNetworks
}

func dfs(_ currentComputer: Int, _ numberOfNetworks: inout Int, _ networks: inout [Int], _ n: Int, _ computers: [[Int]]) {
    if networks[currentComputer] == 0 {
        numberOfNetworks += 1
        networks[currentComputer] = numberOfNetworks
    }
    
    for nextComputer in 0..<n {
        if computers[currentComputer][nextComputer] == 1 && networks[nextComputer] == 0{ 
            networks[nextComputer] = networks[currentComputer]
            dfs(nextComputer, &numberOfNetworks, &networks, n, computers)
        }
    }
}
